lib/extras/gain_map.cc

```diff
@@ -0,0 +1,216 @@
+      4 +                                       // size of compressed_icc_size
+      map_bundle->alt_icc_size +                // size of compressed_icc
+      map_bundle->gain_map_size;                // size of gain map
+  return JXL_TRUE;
+}
+
+JXL_BOOL JxlGainMapWriteBundle(const JxlGainMapBundle* map_bundle,
+                               uint8_t* output_buffer,
+                               size_t output_buffer_size,
+                               size_t* bytes_written) {
+  if (map_bundle == nullptr) return 0;
+
+  uint8_t jhgm_version = map_bundle->jhgm_version;
+
+  FixedSizeMemoryManager<sizeof(jxl::ColorEncoding)> memory_manager;
+  jxl::ColorEncoding internal_color_encoding;
+  jxl::BitWriter color_encoding_writer(memory_manager.memory_manager());
+  if (map_bundle->has_color_encoding) {
+    JXL_RETURN_IF_ERROR(
+        internal_color_encoding.FromExternal(map_bundle->color_encoding));
+    if (!jxl::Bundle::Write(internal_color_encoding, &color_encoding_writer,
+                            /*layer=*/0, nullptr)) {
+      return JXL_FALSE;
+    }
+  }
+
+  color_encoding_writer.ZeroPadToByte();
+
+  size_t cursor = 0;
+  if (cursor + 1 <= output_buffer_size) {
+    memcpy(output_buffer + cursor, &jhgm_version, 1);
+    cursor += 1;
+  }
+
+  uint16_t metadata_size_le = JXL_BSWAP16(map_bundle->gain_map_metadata_size);
+  if (cursor + 2 <= output_buffer_size) {
+    memcpy(output_buffer + cursor, &metadata_size_le, 2);
+    cursor += 2;
+  }
+
+  if (cursor + map_bundle->gain_map_metadata_size <= output_buffer_size) {
+    memcpy(output_buffer + cursor, map_bundle->gain_map_metadata,
+           map_bundle->gain_map_metadata_size);
+    cursor += map_bundle->gain_map_metadata_size;
+  }
+
+  uint8_t color_enc_size =
+      static_cast<uint8_t>(color_encoding_writer.GetSpan().size());
+  if (cursor + 1 <= output_buffer_size) {
+    memcpy(output_buffer + cursor, &color_enc_size, 1);
+    cursor += 1;
+  }
+
+  if (cursor + color_enc_size <= output_buffer_size) {
+    memcpy(output_buffer + cursor, color_encoding_writer.GetSpan().data(),
+           color_enc_size);
+    cursor += color_enc_size;
+  }
+
+  uint32_t icc_size_le = JXL_BSWAP32(map_bundle->alt_icc_size);
+  if (cursor + 4 <= output_buffer_size) {
+    memcpy(output_buffer + cursor, &icc_size_le, 4);
+    cursor += 4;
+  }
+
+  if (cursor + map_bundle->alt_icc_size <= output_buffer_size) {
+    memcpy(output_buffer + cursor, map_bundle->alt_icc,
+           map_bundle->alt_icc_size);
+    cursor += map_bundle->alt_icc_size;
+  }
+
+  if (cursor + map_bundle->gain_map_size <= output_buffer_size) {
+    memcpy(output_buffer + cursor, map_bundle->gain_map,
+           map_bundle->gain_map_size);
+    cursor += map_bundle->gain_map_size;
+  }
+
+  if (bytes_written != nullptr) *bytes_written = cursor;
+  return cursor == output_buffer_size ? JXL_TRUE : JXL_FALSE;
+}
+
+JXL_BOOL JxlGainMapReadBundle(JxlGainMapBundle* map_bundle,
```

Comment:
This function doesn't seem to check that reads are in-bounds, and it definitely should.

Reply:
Added ifs before every read